### PR TITLE
remove usage of Dart_New for paragraph/libtxt

### PR DESCRIPTION
**Pull request by dnfield**
Part of https://github.com/flutter/flutter/issues/50997

TL;DR is that Dart_New is slow and we should try to avoid it.

After this it should just be images.

### CODE REVIEW

File: lib/ui/text.dart [hunk 1]
```diff
@@ -1914,6 +1890,20 @@ class Paragraph extends NativeFieldWrapperClass2 {
   void layout(ParagraphConstraints constraints) => _layout(constraints.width);
   void _layout(double width) native 'Paragraph_layout';
 
+  List<TextBox> _decodeTextBoxes(Float32List encoded) {
+    final List<TextBox> boxes = List<TextBox>(encoded[0].toInt());
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/text.dart]: Done

File: lib/ui/text.dart [hunk 2]
```diff
@@ -1914,6 +1890,20 @@ class Paragraph extends NativeFieldWrapperClass2 {
...
+    for (int index = 0; index < boxes.length; index += 1) {
+      final int position = (index * 5) + 1;
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/text.dart]: Done

File: lib/ui/text.dart [hunk 3]
```diff
@@ -1987,7 +1981,26 @@ class Paragraph extends NativeFieldWrapperClass2 {
   ///
   /// This can potentially return a large amount of data, so it is not recommended
   /// to repeatedly call this. Instead, cache the results.
-  List<LineMetrics> computeLineMetrics() native 'Paragraph_computeLineMetrics';
+  List<LineMetrics> computeLineMetrics() {
+    final Float64List encoded = _computeLineMetrics();
+    final List<LineMetrics> metrics = List<LineMetrics>(encoded[0].toInt());
+    for (int index = 0; index < metrics.length; index += 1) {
+      final int position = (index * 5) + 1;
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/text.dart]: Done, and added some tests that didn't exist that would have caught this.

File: lib/ui/text/paragraph.cc [hunk 4]
```diff
@@ -152,14 +166,34 @@ Dart_Handle Paragraph::getLineBoundary(unsigned offset) {
...
+  // Then there are boxes.size() groups of 9 which are the line metrics
+  // properties
+  auto count = metrics.size() * 9;
+  tonic::Float64List result(
+      Dart_NewTypedData(Dart_TypedData_kFloat64, 1 + count));
+  result[0] = count;
+  for (unsigned long i = 0; i < metrics.size(); i++) {
+    auto position = (i * 9) + 1;
+    const txt::LineMetrics& line = metrics[i];
+    result[position + 0] = static_cast<float>(line.hard_break);
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/text/paragraph.cc]: Done

**Comment by dnfield** [comment]
Also noticed that we could make LineMetrics const constructable.  Not sure how crucial that is since the real use case for it couldn't use it.  At any rate I added hashCode and == and toString overrides for LineMetrics, they were helpful to me in writing tests.

**Comment by jonahwilliams** [comment]
If web has a lineMetrics you should add the logic there too, plus const construction if you decide on it.

**Comment by dnfield** [comment]
@jonahwilliams - strangely, the web `EngineLineMetrics` (implementer of LineMetrics) already has hashCode and equals defined.

I don't want to make this const right now because I'm not sure where it would be useful and it's entirely possible it turns out to be breaking for someone that's constructing it in a non const way where it could have been const.  We can do that in a separate PR if desired.

**Comment by dnfield** [comment]
I'm going to merge this on the idea that the test needs the same logic to pass without these changes - if it's a problem we should fix it separately.

**Pull request merged**